### ISSUE TITLE
Randomness in the optimization progress

### ISSUE THREAD

**Issue by SJoJoK** [graphdeco-inria/gaussian-splatting#89]
Hello,
Thanks for your great work and detailed instructions! I have some confusion when processing my own scenes. Specifically, I'm trying to identify which part of the algorithm (or code) introduce randomness, resulting in fluctuations in performance.
For example, below are 45 results obtained under identical conditions: the same input, parameters, and a consistent random seed.
<img width="649" alt="image" src="https://github.com/graphdeco-inria/gaussian-splatting/assets/50450335/58441ad6-ee2f-4fa3-b02e-612ed7298b76">
The highest PSNR for the test set is 28.06, whereas the lowest stands at 24.23.
From what I've deduced, the algorithmic randomness stems from the **split** in the adaptive control of Gaussians, as it includes sampling. On the code side, randomness appears when "picking a random camera" to render a training view.

Additional Details:
I've made alterations to the `position_lr` (both initial and final), `scaling_lr`, `lambda_dssim`, and `percent_dense` (by the way, I noticed a discrepancy if I didn't go wrong: the default value for `percent_dense` is 0.01, which seems to be inconsistent with the README that mentions it as 0.1).
The scene consists of 20 forward-facing captured indoor images, somehow like LLFF dataset but more complex.
 
Thanks in advance for any help or advise.

**Comment by SJoJoK** [comment]
Information that may help: the results with low PSNR has obvious floaters (precisely, groups of gaussians that are close to the camera's near plane, which overfit to the other training views) in the test views.

**Comment by Snosixtyboo** [comment]
Hi,

thanks for messaging. The randomness of the camera picking and splitting is, in theory, reproducible, since we initialize the random number generator to a known state.  But due to the way that the differentiable rasterizer is implemented, there will always be SOME randomness caused by GPU scheduling. 

If you say that you have floaters, then that is the reason for the score drop. We have no explicit floater control right now other than the alpha resetting because we never needed anything more drastic. Could you let us know how you changed the learning rates and why? We're not sure how far we can take this, but would it be possible to get access to the dataset? We will fix the inconsistency with the README.

Best,
Bernhard

**Comment by SJoJoK** [comment]
Thanks for replying! I agree that the random process is reproducible since they are managed by RNG with the same seed (But according to my previous practice on other projects, the random module of `numpy` and `torch` doesnt work as I expected). However, I'm not familiar with GPU scheduling :(, so I have no opinion on this.

I change the `position_lr` to 4e-6 (initial) to 4e-8 (final), `scaling_lr` to 2e-3, `percent_dense` to 1e-5 (somehow dramatic but it indeed improves score), and `lamda_dssim` to 0.3. The reason I change them is that the default parameters doesnt work well at my scene (PSNR=23.77), so I fine-tune them, and these parameters are determined experimentally.

I'm willing to send you my scene which is already processed for the code. May I send it to your email? 


PS:

I try to control the floater by

* Add near culling during the optimization progress (from viewCoord>0.2 to 5~10)
* Add regularizations (inverse depth smoothness with backward process)
* Clean the outliers of the input sparse point cloud
* Change the reset interval

Only `Clean the outliers of the input sparse point cloud` works, but it introduces other problems.
The result in https://github.com/graphdeco-inria/gaussian-splatting/issues/89#issue-1849057362 only take use of the regularization. But I'm sure that the regularization is deterministic. Also, I'm conducting experiments with the same setting without the regularization, the results are non-deterministic (highest PSNR=27.75, lowest PSNR=24.35)

**Comment by Snosixtyboo** [comment]
Hi,

yes, you can go ahead and send it to my mail (kerbl@cg.tuwien.ac.at), but I am not sure if or how much we can fix this. This may just remain a limitation of the approach for now. The GPU scheduling will cause random behavior, which was not a problem for us, this could only be avoided with a rigorous rewrite.

Best,
Bernhard

**Comment by SJoJoK** [comment]
Sure, I've sent an email with the scene. Thanks again for your help.

**Comment by JIANG-CX** [comment]
> 

Hello, I am curious about the process you use for inverse depth smoothing with the backward approach. Can you please elaborate on it?

**Comment by SJoJoK** [comment]
> > 
> 
> Hello, I am curious about the process you use for inverse depth smoothing with the backward approach. Can you please elaborate on it?

Sure, you may check out this link https://kornia.readthedocs.io/en/latest/losses.html#kornia.losses.InverseDepthSmoothnessLoss for more details.

**Comment by JIANG-CX** [comment]
> > > 
> > 
> > 
> > Hello, I am curious about the process you use for inverse depth smoothing with the backward approach. Can you please elaborate on it?
> 
> Sure, you may check out this link https://kornia.readthedocs.io/en/latest/losses.html#kornia.losses.InverseDepthSmoothnessLoss for more details.

Thank you for your response. Can you please provide more information on how to obtain the depth image? My concern is that simply splatting the depth of the ellipsoid's center onto the image plane as how the SH does may not be accurate, as the depth varies across different areas of the ellipsoid.

**Comment by SJoJoK** [comment]
> > > > 
> > > 
> > > 
> > > Hello, I am curious about the process you use for inverse depth smoothing with the backward approach. Can you please elaborate on it?
> > 
> > 
> > Sure, you may check out this link https://kornia.readthedocs.io/en/latest/losses.html#kornia.losses.InverseDepthSmoothnessLoss for more details.
> 
> Thank you for your response. Can you please provide more information on how to obtain the depth image? My concern is that simply splatting the depth of the ellipsoid's center onto the image plane as how the SH does may not be accurate, as the depth varies across different areas of the ellipsoid.

I get the depth map by just simply changing the alpha-blending process from $\sum T \alpha c$ to $\sum T \alpha d$, where d is the depth. This is indeed an approximation, but is useful for visualization/regularization.

**Comment by JIANG-CX** [comment]
> > > > > 
> > > > 
> > > > 
> > > > Hello, I am curious about the process you use for inverse depth smoothing with the backward approach. Can you please elaborate on it?
> > > 
> > > 
> > > Sure, you may check out this link https://kornia.readthedocs.io/en/latest/losses.html#kornia.losses.InverseDepthSmoothnessLoss for more details.
> > 
> > 
> > Thank you for your response. Can you please provide more information on how to obtain the depth image? My concern is that simply splatting the depth of the ellipsoid's center onto the image plane as how the SH does may not be accurate, as the depth varies across different areas of the ellipsoid.
> 
> I get the depth map by just simply changing the alpha-blending process from ∑Tαc to ∑Tαd, where d is the depth. This is indeed an approximation, but is useful for visualization/regularization.

Thank you for your feedback. However, I have noticed that there are instances where there are only a few points with a large covariance on a uniformly colored floor or wall. It is possible that using this method in such areas could result in significant errors.

**Comment by SJoJoK** [comment]
> > > > > > 
> > > > > 
> > > > > 
> > > > > Hello, I am curious about the process you use for inverse depth smoothing with the backward approach. Can you please elaborate on it?
> > > > 
> > > > 
> > > > Sure, you may check out this link https://kornia.readthedocs.io/en/latest/losses.html#kornia.losses.InverseDepthSmoothnessLoss for more details.
> > > 
> > > 
> > > Thank you for your response. Can you please provide more information on how to obtain the depth image? My concern is that simply splatting the depth of the ellipsoid's center onto the image plane as how the SH does may not be accurate, as the depth varies across different areas of the ellipsoid.
> > 
> > 
> > I get the depth map by just simply changing the alpha-blending process from ∑Tαc to ∑Tαd, where d is the depth. This is indeed an approximation, but is useful for visualization/regularization.
> 
> Thank you for your feedback. However, I have noticed that there are instances where there are only a few points with a large covariance on a uniformly colored floor or wall. It is possible that using this method in such areas could result in significant errors.

Thanks for the information, I also notice the under-reconstruction of texture-less plane, and yet I have no idea of tackling this:(

**Comment by JIANG-CX** [comment]
> > > > > > > 
> > > > > > 
> > > > > > 
> > > > > > Hello, I am curious about the process you use for inverse depth smoothing with the backward approach. Can you please elaborate on it?
> > > > > 
> > > > > 
> > > > > Sure, you may check out this link https://kornia.readthedocs.io/en/latest/losses.html#kornia.losses.InverseDepthSmoothnessLoss for more details.
> > > > 
> > > > 
> > > > Thank you for your response. Can you please provide more information on how to obtain the depth image? My concern is that simply splatting the depth of the ellipsoid's center onto the image plane as how the SH does may not be accurate, as the depth varies across different areas of the ellipsoid.
> > > 
> > > 
> > > I get the depth map by just simply changing the alpha-blending process from ∑Tαc to ∑Tαd, where d is the depth. This is indeed an approximation, but is useful for visualization/regularization.
> > 
> > 
> > Thank you for your feedback. However, I have noticed that there are instances where there are only a few points with a large covariance on a uniformly colored floor or wall. It is possible that using this method in such areas could result in significant errors.
> 
> Thanks for the information, I also notice the under-reconstruction of texture-less plane, and yet I have no idea of tackling this:(

OK, Thanks.

**Comment by JIANG-CX** [comment]
> Sure, I've sent an email with the scene. Thanks again for your help.

Hello, I was wondering if you have found a solution to the issue of randomness?

**Comment by ricshaw** [comment]
Any update on the cause of randomness? I get different results every time I train, even with random seeds set...

**Comment by ricshaw** [comment]
This is still a major issue. Has anyone looked into this?

**Comment by BitCalSaul** [comment]
> This is still a major issue. Has anyone looked into this?

Here, too.

**Comment by BitCalSaul** [comment]
Any updates?
